### DISCUSSION TITLE
Receiving CORS option redirect error even with CORS headers set

### DISCUSSION THREAD

**Discussion by scottsandersdev** [mockoon/mockoon#619] [Help]
I've setup a mock api to return a simple object, and it's working fine when hitting it with postman, but when I try to make the request from my webapp, I get an error 
```
Access to XMLHttpRequest at '' from origin '' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.
```
I've tried using the "Automatically handle OPTIONS pre-flight requests" setting as well as creating my own options route that returns 200 and for some reason Chrome still seems to think that the options request is attempting to redirect.

Any ideas on solutions?

**Screenshots**
![Screen Shot 2021-12-10 at 12 02 11 PM](https://user-images.githubusercontent.com/1984888/145612982-d7cefba9-eedf-4db6-b70e-111b1461f9cf.png)

**Mockoon version:**
1.17.0

**OS / OS version:**
Mac OS

**Comment by 255kb** [comment]
Could you please share the HTTP transaction URL, status, request and response headers, etc.? 
When you use the "automatically handle pre-flight requests" option, Mockoon will always answer with status 200 to OPTIONS requests, and add the following headers: 

```http
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET,POST,PUT,PATCH,DELETE,HEAD,OPTIONS'
Access-Control-Allow-Headers: Content-Type, Origin, Accept, Authorization, Content-Length, X-Requested-With
```

But it seems that instead Mockoon (or the proxied server?) replied with a redirection.

**Reply by scottsandersdev** [reply]
Chrome devtools isn't showing much due to the error, but I'll post what I have below. I also have some screenshots from postman, which shows the correct headers from mockoon.

Request Headers in chrome - 
![Screen Shot 2021-12-15 at 12 10 42 PM](https://user-images.githubusercontent.com/1984888/146234081-80d68886-040c-43ea-8624-143df24a182a.png)

Response Headers in chrome - 
![Screen Shot 2021-12-15 at 12 12 21 PM](https://user-images.githubusercontent.com/1984888/146234111-20d0d0ff-e030-4e44-b4a2-a9b487b7956e.png)

Postman options request - 
![Screen Shot 2021-12-15 at 12 14 13 PM](https://user-images.githubusercontent.com/1984888/146234175-320000bf-addb-4fde-b415-543232cac2f2.png)

**Reply by 255kb** [reply]
This is rather strange, I never encountered such issue. Could you maybe share your environment (using the export menu)? I could try to reproduce.

**Reply by scottsandersdev** [reply]
I created a repo with a react app that attempts to hit the endpoints in mockoon. The repo also has the environment json in it (mockoon-test-api.json). Once you have mockoon running just cd into the project folder and then run npm i && npm start. 
https://github.com/scottsandersdev/mockoon-test. 
All of the endpoints in mockoon are returning cors errors for me when fetched through chrome.

**Reply by 255kb** [reply]
I just tried to run your application and to set up a fake domain `dev.local` pointing to `localhost` like you did. 
All the requests I tried triggered OPTION pre-flight requests and Mockoon answered them correctly. 
I added one more DELETE route just to test: 
![image](https://user-images.githubusercontent.com/7489814/146787041-73939ba1-f865-4e65-93a2-16ff25c4d312.png)

Maybe you have a configuration somewhere, like in the hosts file, that makes your `localhost:1236` redirects somehow?

**Reply by scottsandersdev** [reply]
Interesting. This is what my hosts file looks like - 
```
##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1	localhost dev.local
255.255.255.255	broadcasthost
::1             localhost
# Added by Docker Desktop
# To allow the same kube context to work on the host and the container:
127.0.0.1 kubernetes.docker.internal
# End of section
```
I don't see anything suspicious. Any ideas where else I could look for issues?

**Reply by 255kb** [reply]
Me neither. I have honestly no clue where this could come from. What I found here and there on the internet were mainly issues with redirect configuration from `/path` to `/path/` in Apache for example. But there is no such redirect logic anywhere in Mockoon. The CORS option only answer with a 200 to OPTIONS requests (except for the OPTIONS you declare yourself). 

What I would try to check: 
- you don't have a `Location` in your Mockoon's global headers. But I doubt it as you would have issues with other routes/methods too.
- there is no external tool like Apache/nginx that could interfere with or proxy to Mockoon's API and causing a redirect

**Reply by scottsandersdev** [reply]
Little bit of an update here. I tried playing with this some more today because I'd really like to recommend this tool to my team. I noticed that if I opened an incognito browser in chrome and then go to the Headers tab in Mockoon and "add CORS headers" that the requests started working. Note that I had "Automatically handle Options pre-flight requests" checked before adding those CORS headers in the Headers tab.

If I'm not running incognito this is what I see, even with CORS headers added. 

<img width="1197" alt="Screen Shot 2022-01-20 at 6 01 00 PM" src="https://user-images.githubusercontent.com/1984888/150435722-a8205c89-8283-41df-9aab-e8fd635fc6c7.png">

Most of these are net::ERR_SSL_PROTOCOL_ERROR, though the options request returns a cors error.

So my questions are - Is there a way to make this work with mockoon without using an incognito window? Does mockoon attempt to upgrade http to https? If so, is there a way to disable that? Any other thoughts on what the issue is?

**Reply by 255kb** [reply]
It looks like your `http://` calls are automatically redirected to `https://`.
Could it be a browser cache problem? https://superuser.com/questions/565409/how-to-stop-an-automatic-redirect-from-http-to-https-in-chrome

To answer your other question Mockoon will never use TLS/HTTPS if this option is not enabled in the environment settings (last header tab). I see in your [mock](https://github.com/scottsandersdev/mockoon-test/blob/main/mockoon-test-api.json#L94-L102) that TLS is disabled so it should be ok.

**Reply by scottsandersdev** [reply]
Unfortunately the hsts trick didn't change anything. I'd tried that before, but went ahead and tried again for good measure and still seeing the same result. Just to confirm - when you ran my code locally were you in incognito mode or no?

**Reply by 255kb** [reply]
I tested your app and mock without using incognito mode.

**Reply by scottsandersdev** [reply]
I see. I'll go ahead and create a custom express mock api. Thanks for all the quick replies.

**Reply by 255kb** [reply]
Sorry that we couldn't solve your issue! I am out of ideas 🤔
